src/coreclr/jit/gcinfo.cpp

```diff
@@ -252,8 +252,23 @@ GCInfo::WriteBarrierForm GCInfo::gcIsWriteBarrierCandidate(GenTreeStoreInd* stor
         return WBF_NoBarrier;
     }
 
+    // Write-barriers are no-op for frozen objects (as values)
     if (store->Data()->IsIconHandle(GTF_ICON_OBJ_HDL))
     {
+#ifndef TARGET_XARCH
+        const ssize_t handle = store->Data()->AsIntCon()->IconValue();
+        if (!compiler->info.compCompHnd->isObjectImmutable(reinterpret_cast<void*>(handle)))
+        {
+            // On platforms with weaker memory model we need to make sure we use a store with the release semantic
+            // when we publish a potentially mutable object
+            // See relevant discussions https://github.com/dotnet/runtime/pull/76135#issuecomment-1257258310 and
+            // https://github.com/dotnet/runtime/pull/76112#discussion_r980639782
+
+            // This can be relaxed to "just make sure to use stlr/memory barrier" if needed
+            store->gtFlags |= GTF_IND_VOLATILE;
```

Comment:
> `store->gtFlags |= GTF_IND_VOLATILE;`

It does not look right to have this idempotent function mutate the underlying store. It also leads to 2 unnecessary VM calls (in LSRA / codegen), so I think it would be better to mark the store once, in lowering, unless there's some problem with that.

Reply:
@SingleAccretion what's the problem with the current code - it looks like the best place to decide whether write barrier is needed or not, yet it adds a flag to a store which look kinda bad but at least it's clear in what cases we get rid of WB.

Reply:
> 2 unnecessary VM calls (in LSRA / codegen),

I am not sure I understand which vm calls you refer to?

Reply:
Ah, you meant the flag only, and the `if (store->Data()->IsIconHandle(GTF_ICON_OBJ_HDL))` check is fine, right? then yes, we can move it I assume

Reply:
> I am not sure I understand which vm calls you refer to?

`gcIsWriteBarrierCandidate` is called 3 times for a given store:

1) `LowerStoreIndirCommon` (lower).
2) `BuildNode` (LSRA).
3) `genCodeForStoreInd` (codegen).

> but at least it's clear in which cases we get rid of WB in just one place.

I'd say the fact we have a write barrier vs need the release semantics is a bit orthogonal. It "just so happens" the write barrier provides them and we don't need to be explicit about it. You can still have the logic centralized in place though, by adding a `bool* pExplicitReleaseNeeded = nullptr` sort of parameter to this method, then lowering would query it to decide whether `GTF_IND_VOLATILE` needs to be applied, and other phases not.

Reply:
Thanks, totally makes sense, moved